src/field_theory/finite/chevalley_warning.lean

```diff
@@ -0,0 +1,176 @@
+        { intros x hx, rw mem_filter at hx ⊢, exact ⟨mem_univ _, if_pos rfl⟩ },
+        { intros x hx, apply prod_congr rfl, intros j hj, rw if_neg, rintro rfl, exact hi₀ hj },
+        { intros x y hx hy hxy, rw mem_filter at hx hy, rw function.funext_iff at hxy ⊢,
+          intros j, specialize hxy j,
+          by_cases hj : j = i₀, { rw [hj, hx.2, hy.2] }, { rwa [if_neg hj, if_neg hj] at hxy } },
+        { intros x hx, refine ⟨λ j, if j = i₀ then a else x j, _, _⟩,
+          { rw mem_filter at hx ⊢, exact ⟨mem_univ _, if_pos rfl⟩ },
+          { funext j, split_ifs with hj, { rw mem_filter at hx, rw [hj, hx.2] }, { refl } } } },
+      { intros x hx y hy hxy, rw disjoint_iff, contrapose! hxy,
+        rcases exists_mem_of_ne_empty hxy with ⟨z, hz⟩,
+        rw [inf_eq_inter, mem_inter, mem_filter, mem_filter] at hz,
+        rcases hz with ⟨⟨_, rfl⟩, ⟨_, rfl⟩⟩, refl } },
+    { rw eq_univ_iff_forall, intro x, rw mem_bind,
+      refine ⟨x i₀, mem_univ _, _⟩, rw mem_filter, exact ⟨mem_univ _, rfl⟩ } },
+end
+
+/-- The Chevalley–Warning theorem.
+   Let (f i) be a finite family of multivariate polynomials
+   in finitely many variables (X s, s : σ)
+   over a finite field of characteristic p.
+   Assume that the sum of the total degrees of the f i is less than the cardinality of σ.
+   Then the number of common solutions of the f i is divisible by p. -/
+theorem char_dvd_card_solutions (p : nat.primes) [char_p K p]
+  {ι : Type*} [decidable_eq ι] (s : finset ι) (f : ι → (mv_polynomial σ K))
+  (h : (s.sum $ λ i, (f i).total_degree) < fintype.card σ) :
+  (p:ℕ) ∣ fintype.card {x : σ → K // ∀ i ∈ s, (f i).eval x = 0} :=
+begin
+  have hq : 0 < q - 1,
+  { rw [← card_units, fintype.card_pos_iff],
+    exact ⟨1⟩ },
+  let F : mv_polynomial σ K := s.prod (λ i, (1 - (f i)^(q-1))),
+  suffices : univ.sum (λ x, F.eval x) =
+    fintype.card {x : σ → K // ∀ i ∈ s, (f i).eval x = 0},
+  { rw [← char_p.cast_eq_zero_iff K, ← this],
+    apply sum_mv_polynomial_eq_zero,
+    calc F.total_degree ≤ s.sum (λ i, (1 - (f i)^(q-1)).total_degree) :
+      total_degree_finset_prod s _
+      ... ≤ s.sum (λ i, (q - 1) * (f i).total_degree) :
+      begin
+        apply sum_le_sum,
+        intros i hi,
+        refine le_trans (total_degree_sub _ _)
+          (le_trans _ (total_degree_pow _ _)),
+        simp only [max_eq_right, nat.zero_le, total_degree_one]
+      end
+      ... = (q - 1) * (s.sum $ λ i, (f i).total_degree) : mul_sum.symm
+      ... < (q - 1) * (fintype.card σ) : by rwa mul_lt_mul_left hq },
+  { let S : finset (σ → K) := univ.filter (λ x : σ → K, ∀ i ∈ s, (f i).eval x = 0),
+    rw [fintype.card_of_subtype S, card_eq_sum_ones, nat_cast_sum, nat.cast_one,
+     ← fintype.sum_extend_by_zero S],
+    { apply sum_congr rfl,
+      intros x hx, clear hx,
+      rw show F.eval x = finset.prod s (λ (i : ι), (1 - f i ^ (q - 1)).eval x),
+      { convert eval₂_prod _ _ _ _, exact is_semiring_hom.id },
+      split_ifs with hx hx,
```

Comment:
You could make the rest of this proof cleaner by defining a lemma `a ^ (q - 1) = if a = 0 then 0 else 1`, then rewriting inside the product with `conv` (before doing `split_ifs`), then writing a lemma for multiplying a bunch of `if p i then 0 else 1`s together, if it doesn't already exist.

Reply:
Still a TODO, but Github thinks it's "outdated" because the code moved.

Reply:
Done